Fix Issue #31 TagMacro: resolve parameters inside type bounds & wildcards

cmofx opened the pull request:
/claim #31  

I added new function wildcardType in LightTypeTag to create bounds wildcard type tag.

Also changed the TacMacro in scala 3 and scala 2 to parse the tag type bound parameters and summon the type tags accordingly 

Also fixed a bug in the scala 2 ReflectionUtil.scala that TypeBound's lower bound is not checked for strong type

cmofx committed:
Fix Issue #31

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zio/izumi-reflect?pullRequest=537) <br/>All committers have signed the CLA.

neko-kai commented:
@cmofx Great work! :+1:

Could you please also tests for recursive bound case, such as `x[T[_]: TagK, U: Tag] = Tag[Set[_ <: T[U]]]` ?

cmofx committed:
Added the support for recursive bound case

cmofx commented:
> @cmofx Great work! 👍
> 
> Could you please also tests for recursive bound case, such as `x[T[_]: TagK, U: Tag] = Tag[Set[_ <: T[U]]]` ?

Just added the support for recursive bound case, updated the SharedTagTest.scala

cmofx committed:
Fixed one compiling error

pshirshov commented:
Looks nice for me, thank you for your help.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala/izumi/reflect/macrortti/LightTypeTag.scala:
```diff
@@ -353,6 +353,18 @@ object LightTypeTag {
   }
 
+  def wildcardType(lowTag: LightTypeTag, highTag: LightTypeTag): LightTypeTag = {
+    val ref = LightTypeTagRef.WildcardReference(Boundaries.Defined(lowTag.ref.asInstanceOf[AbstractReference], highTag.ref.asInstanceOf[AbstractReference]))
```
```suggestion
    val ref = LightTypeTagRef.WildcardReference(Boundaries.Defined(lowTag.ref match { case r: AbstractReference => r }, highTag.ref match { case r: AbstractReference => r }))
```
As in LightTypeTag.refinedType above

neko-kai reviewed (approved):
LGTM, will try to merge tomorrow :+1:

cmofx committed:
Updated the way to map LightTypeTag.ref to AbstractReference

cmofx commented:
Updated the way to map LightTypeTag.ref to AbstractReference

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -56,6 +56,13 @@ private[reflect] object ReflectionUtil {
+      case e: Universe#ExistentialTypeApi =>
+        allPartsStrong(outerTypeParams, e.underlying) &&
+          e.underlying.typeArgs.forall(t => t.typeSymbol.typeSignature match {
+            case tb: Universe#TypeBounds => allPartsStrong(outerTypeParams, tb.hi) && allPartsStrong(outerTypeParams, tb.lo)
```
```suggestion
            case tb: Universe#TypeBoundsApi => allPartsStrong(outerTypeParams, tb.hi) && allPartsStrong(outerTypeParams, tb.lo)
```
To avoid warning `abstract type pattern scala.reflect.api.Universe#TypeBounds is unchecked since it is eliminated by erasure`

neko-kai commented on izumi-reflect/izumi-reflect/src/test/scala/izumi/reflect/test/SharedTagTest.scala:
```diff
@@ -1093,6 +1093,22 @@ abstract class SharedTagTest extends AnyWordSpec with XY[String] with TagAsserti
+      // for recursive bound case
+      def x[T[_]: TagK  , U: Tag] = Tag[Set[_ <: T[U]]]
+      assertRepr(x[List, Long].tag, "Set[=?: <Nothing..List[+Long]>]")
+      assertSame(x[List, Long].tag, Tag[Set[_ <: List[Long]]].tag)
```
@cmofx Given the following additional tests:

```suggestion
      assertSameStrict(x[List, Long].tag, Tag[Set[_ <: List[Long]]].tag)

      def xcov[T[_]: TagK, U: Tag] = Tag[List[_ <: T[U]]]
      assertRepr(xcov[List, Long].tag, "List[+?: <Nothing..List[+Long]>]")
      assertSameStrict(xcov[List, Long].tag, Tag[List[_ <: List[Long]]].tag)

      def xcontr[T[_]: TagK, U: Tag] = Tag[(_ <: T[U]) => Int]
      assertRepr(xcontr[List, Long].tag, "Function1[-?: <Nothing..List[+Long]>,+Int]")
      assertSameStrict(xcontr[List, Long].tag, Tag[(_ <: List[Long]) => Int].tag)
```

`xcov` case fails with
```
"List[+[List[+Long]]]" did not equal "List[+[?: <Nothing..List[+Long]>]]" (TagAssertions.scala:27)
```
Only on Scala versions 2.13 and 2.12 and works on Scala 2.11 and 3.3

Could you try to investigate whether this can be fixed?

cmofx replied:
The xcov failed because the type bounds information is lost in
src/main/scala-2/izumi/reflect/TagMacro.scala, line 106, 
    val tgt = ReflectionUtil.norm(c.universe: c.universe.type, logger)(tpe.dealias)
where tpe.dealias removed the type bounds 

As dealias is implemented internally, I think it is not fixable.

neko-kai replied:
Yes, that does not seem to be reasonably fixable

neko-kai reviewed (commented):
@cmofx I reviewed in detail and found a failure when TagMacro combination inside bounds is combined with covariance on Scala 2.12/2.13, could you please look at it? Thanks. The details are in the PR comments

cmofx committed:
Changed TypeBounds to TypeBoundsApi
